### ISSUE TITLE
How to make wsl2 alive in the background

### ISSUE THREAD

**Issue by amunv** [microsoft/WSL#10138]
Wsl2 will automatically shut down when the terminal is closed.I need it alive in the background.
Is there such a function in the configuration file, so that wsl2 can only be closed manually by running wsl --shutdown. If not, would you consider adding this feature.

**Comment by elsaco** [comment]
@amunv WT has a headless feature (https://github.com/microsoft/terminal/pull/14944) that might keep your WSL instance running.

**Comment by amunv** [comment]
> @amunv WT has a headless feature ([microsoft/terminal#14944](https://github.com/microsoft/terminal/pull/14944)) that might keep your WSL instance running.

I don't understand it a bit, should I add "allowHeadless": true in setting.json
As he said, there is no wt --headless or wt --hidden.

**Comment by aki-k** [comment]
@amunv You can make it stay running with this setup: https://github.com/microsoft/WSL/issues/10157#issuecomment-1578024408

**Comment by OneBlue** [comment]
Thank you for reporting this @amunv. You can force WSL2 to keep running with: 

```
[wsl2]
vmIdleTimeout=-1
```

in %userprofile%/.wslconfig

**Comment by aki-k** [comment]
> vmIdleTimeout=-1

I'm not sure why this is not documented at https://learn.microsoft.com/en-us/windows/wsl/wsl-config

It just says

> The number of milliseconds that a VM is idle, before it is shut down. Only available for Windows 11.

**Comment by amunv** [comment]
> Thank you for reporting this @amunv. You can force WSL2 to keep running with:
> 
> ```
> [wsl2]
> vmIdleTimeout=-1
> ```
> 
> in %userprofile%/.wslconfig

This is not what I need. About ten seconds after closing the terminal, wsl enters the suspended state. At this time, the docker application cannot be used normally, and the system memory usage is reduced. The "vmIdleTimeout" option controls the survival time of vmmemwsl after wsl is suspended. During this time, when WT is turned on, wsl will run normally. I don't need this suspend state and control the time after suspend. What I need is that wsl suspend after about ten seconds, disable this suspend.
`[wsl2]
vmIdleTimeout=120000
memory=2GB 
processors=2
`

**Comment by OneBlue** [comment]
> > Thank you for reporting this @amunv. You can force WSL2 to keep running with:
> > ```
> > [wsl2]
> > vmIdleTimeout=-1
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > in %userprofile%/.wslconfig
> 
> This is not what I need. About ten seconds after closing the terminal, wsl enters the suspended state. At this time, the docker application cannot be used normally, and the system memory usage is reduced. The "vmIdleTimeout" option controls the survival time of vmmemwsl after wsl is suspended. During this time, when WT is turned on, wsl will run normally. I don't need this suspend state and control the time after suspend. What I need is that wsl suspend after about ten seconds, disable this suspend. `[wsl2] vmIdleTimeout=120000 memory=2GB processors=2 `

Yes, that's exactly what `vmIdleTimeout=-1` does. When set to -1, the WSL2 VM will never time out and WSL2 will keep running until wsl.exe --shutdown is called.

**Comment by elsaco** [comment]
@aki-k `vmIdleTimeout = -1` does keep the VM running after exiting the shell. You stated adding the setting then `restarted WSL 2 VM`. Make sure you restart the WSL service not only the VM. Run `wsl --shutdown` for the setting to take effect.

**Comment by aki-k** [comment]
> and restarted WSL 2 VM.

means I exited the WSL 2 VM and ran wsl --shutdown in CMD prompt. Do I also need to restart lxssmanager?

**Comment by aki-k** [comment]
After "exit" -> 2 wslhost.exe processes stop -> "bash", I can use "ps auwwwf | grep apache2" and they just started up. So the WSL 2 instance also started up again.

**Comment by aki-k** [comment]
I actually now found a command that starts WSL 2 VM, exits, and WSL 2 VM and instance are left running:
```
wsl --exec dbus-launch true
```

**Comment by amunv** [comment]
> I actually now found a command that starts WSL 2 VM, exits, and WSL 2 VM and instance are left running:
> 
> ```
> wsl --exec dbus-launch true
> ```

This command is very useful, but I have another problem about WT.The command line of Profiles is run by cmd in windows terminal.The exec parameter cannot be recognized normally in cmd, but it can be recognized in powershell.Do you have an idea to change the command line to pwsh, or let cmd recognize "-e"?

**Comment by aki-k** [comment]
@amunv 

> The exec parameter cannot be recognized normally in cmd, but it can be recognized in powershell.

I don't use Windows Terminal, just CMD prompt (with clink) and it works in it.

**Comment by amunv** [comment]
> @amunv
> 
> > The exec parameter cannot be recognized normally in cmd, but it can be recognized in powershell.
> 
> I don't use Windows Terminal, just CMD prompt (with clink) and it works in it.

Sorry,I just misunderstood this command, I thought this would start the wsl instance.

**Comment by aki-k** [comment]
@amunv Yes, it starts the WSL 2 VM and the WSL 2 instance.

**Comment by aki-k** [comment]
@amunv i.e.
```
C:\Users\aki>wsl --exec dbus-launch true

C:\Users\aki>bash
```

**Comment by BarathwajAnandan** [comment]
> I actually now found a command that starts WSL 2 VM, exits, and WSL 2 VM and instance are left running:
> 
> ```
> wsl --exec dbus-launch true
> ```


Savior for real. In most cases the timeout = -1 or dbus-launch true should work.

for context, if anyone cares ,I am trying to run a ssh-server to login into WSL2 from my macbook. This was the final piece of the puzzle <3. Thanks much.

**Comment by Ali-thepro** [comment]
@aki-k 
wsl --exec dbus-launch true

This works for me, the wsl stays running. But After a reboot do you need to put the command in again? After I rebooted it said it was running but after a minute it said it stopped.

**Comment by MRBZ-GIT** [comment]
> I actually now found a command that starts WSL 2 VM, exits, and WSL 2 VM and instance are left running:
> 
> ```
> wsl --exec dbus-launch true
> ```

For Windows 10 (because the `vmIdleTimeout = -1` method only works for Windows 11), we can further refine @aki-k's method above to make the process automatic by adding the following bash commands to your wsl distro, e.g., Ubuntu's `.bashrc` file:
```
#-----------------------
# keep wsl distro alive
#-----------------------
if ! pgrep -x "dbus-daemon" > /dev/null; then
   dbus-launch true
fi
```
The `pgrep` line is to ensure that we don't create multiple dbus processes for this purpose (1 process suffices).

**Comment by nodesocket** [comment]
Running Windows 11 Pro, and my `.wslconfig` is:

```
[wsl2]
vmIdleTimeout=-1
memory=8GB
processors=4
swap=0
```

However, when I close the WSL Ubuntu window my systemd services stop. Is there a way to continue to have them run when the Ubuntu window is closed?

**Comment by aki-k** [comment]
@nodesocket The solution (until Microsoft fixes it) is to start WSL 2 like this:

C:\\>wsl.exe --distribution Ubuntu-22.04 --exec dbus-launch true & wsl.exe --distribution Ubuntu-22.04

**Comment by nodesocket** [comment]
> @nodesocket The solution (until Microsoft fixes it) is to start WSL 2 like this:
> 
> C:\>wsl.exe --distribution Ubuntu-22.04 --exec dbus-launch true & wsl.exe --distribution Ubuntu-22.04

Thanks for the reply. I'm a macOS and Linux guy, is there a way to make the above a shortcut?

**Comment by aki-k** [comment]
@nodesocket Well try this :)
```
@echo off

wsl.exe --distribution Ubuntu-22.04 --exec dbus-launch true & wsl.exe --distribution Ubuntu-22.04
```

Put the above in start-wsl-2.bat in your %PATH% / $env:PATH

**Comment by aki-k** [comment]
@nodesocket Here's a better solution that doesn't open an extra CMD prompt/console window:
```
Set objShell = WScript.CreateObject("WScript.Shell")
objShell.Run "wsl.exe --distribution Ubuntu-22.04 --exec dbus-launch true", 0, True
objShell.Run "wsl.exe --distribution Ubuntu-22.04", 1, True
```
^- start-wsl-2.vbs

**Comment by fliespl** [comment]
I have added there lines to .bashrc:


```
# Check if DBus is running, if not, launch it
if ! pgrep -x "dbus-launch" > /dev/null; then
    eval `dbus-launch --sh-syntax`
fi
```

**Comment by gardengim** [comment]
I just write a simple systemd service keeping WSL instance alive: [keepwsl](https://github.com/gardengim/keepwsl)

**Comment by orwithout** [comment]
在 ~/.bashrc 中添加 dbus-daemon 进程保活：
> pgrep -u "$(whoami)" -x "dbus-daemon" >/dev/null || dbus-launch true &>/dev/null  #keep_wsl_distro_alive

当 wsl --shutdown 后再次进入子系统实例，仍然会出现两个 dbus-daemon 进程：
![Image](https://github.com/user-attachments/assets/8a8ec967-c482-422c-915c-19bafb48da5c)

我愿意改为在 ~/.bashrc 中添加：
> pgrep -f "keepalive-sleep" > /dev/null || nohup bash -c 'exec -a keepalive-sleep sleep infinity' &>/dev/null &

这样只需要登录wsl 则会自动保活.
如果要开启"宿主windows系统登录便自动生效"，则在 shell:startup 中创建一个vbs 脚本，写入：

> Dim wsl_name : wsl_name = "Ubuntu"
> WScript.CreateObject("WScript.Shell").Run "wsl.exe -d " & wsl_name , 1, True

**Comment by aki-k** [comment]
> My solution right now until MicroSoft add that as an option

Microsoft designed the WSL 2 instance to shutdown automatically.

**Comment by JGu-K** [comment]
The method works on distributions using dbus-broker as the systemd dbus service.
However wouldn't running both dbus-broker and dbus-daemon at the same time cause an error?

I don't understand why dbus-broker doesn't work the same way. Is it a mistake in Microsoft's implementation?

**Comment by gestate-8** [comment]
The method works on distributions using dbus-broker as systemd dbus service.
However wouldn't running both dbus-broker and dbus-daemon at the same time cause an error?

I don't understand why dbus-broker doesn't work the same way. Is it a mistake in Microsoft's implementation?

**Comment by benson-yeh** [comment]
About the WSL2 auto-shutdown issue, honestly, WSL2 is supposed to be a tool for developers. I can't imagine any developer would want it to automatically shut down. It’s super frustrating and messes up your workflow, especially when you’re in the middle of something. This issue has been around for almost two years, and it still hasn’t been addressed. Sure, some people have suggested workarounds, but this really is something Microsoft should be handling, not the community. It really shouldn’t be happening with a tool meant to help with development. Hopefully, they fix this soon, but it’s honestly disappointing that it’s still an issue.

**Comment by randomseed42** [comment]
I've added below script in .bashrc and it works fine for now:

```sh
# keep wsl alive
if [[ $- == *i* ]] && ! pgrep -f "sleep infinity" > /dev/null; then
  nohup sleep infinity >/dev/null 2>&1 &
fi
```

**Comment by Adhjie** [comment]
Yeah, auto-shutdown is weird. If WSL is made to be server-first, and for devs then, server auto-shutdown without admin permission is, safe to say. the nightmare of any devs (critical sever in production going offline...out of scheduled maintenance...), so it's weird, what's the decision that led to auto-shutdown anyway.
there's a WSL GUI settings as windows app, it should be as easy as, tab saying:
server mode (critical):no shutdown (dpkg-divert, idk how to suspend in apt, snap refresh --hold), I think a better analogy of this auto-shutdown is, if android has auto task-killer.
scheduled shutdown option: same as timeout but should also have:
scheduled restart option:
There's a similar dilemma in docker desktop in windows by docker, wsl only feature of mirrored network and loop back host address, that hinders a possibility to just install and use docker exclusively in wsl distro because interaction between windows and wsl is an overhead.
It's confirmed to be a paravirtualization in XWayland workaround for WSL.
Alright I guess, it's a dependency problem (a mild case of dependency hecc).
linux no shutdown but dual boot.
wsl + linux distro, 2 dependencies adding micro soft as a factor.
in any case, this should be a priority since it's a normal state for server mode linux, which most devs use, with WSL GUI being even more unexplored.

even a temporary fix of including wsl.config with vmIdleTimeout = (24 hours in miliseconds) is made after installing wsl, and then a task in task scheduler launching wsl in terminal every 23 hours 59 mins is made after that as dependency should work.

sigh, I guess. the fireship or fireship comment section, any other CS vids saying ring true (this is an old joke, I just checked it in other forums, but weirdly no direct form of this in xcd panel, most xcd only cover the technical debt concept):
Nothing is more permanent than a temporary solution
https://custom-images.strikinglycdn.com/res/hrscywv4p/image/upload/c_limit,fl_lossy,h_9000,w_1200,f_auto,q_auto/784839/2271_379433.jpeg

Sorry, if this rant gets into your email. I just realized this would be a bigger spaghetti code, or technical debt if left not fixed. I just get this problem in brave too, so the last comment about linux server mode should be manual shutdown really trigger this text. if this gets into email's notification*, just deleted it.

tldr
(this is probably the worst case of "If ain't broke dont fix it" but its about the workaround and not the main software.
this one step backward from failsafe, where there are now 2 points of failure, a code that broke this, and a code that broke the workaround.)
technical debt for dev, no failsafe for user. (all in all not really fun, it reminds me of the npm repo deletion case, and the famous SSL case.
I guess manual save, just like in word instead of auto save, because wsl could shutdown without warning, until this get fix by ms hopefully.)

(this reminds me that there's an option to cancel hibernate in gaming android emulator in pc, so that one has more feature, is it the incentive problem?
IK, a software where its good, because the dev also use it as a daily drive, so is the problem, ms also dont use it? test in prod only?)

**Comment by vbkukshaus** [comment]
Hi Folks,

not sure, if the issue is still up-to-date or not. 
I've solved that on my system by creating a new task within the **Task Scheduler**:

Just create a new task, call it "WSL" or something like that, and do the following settings:

On the first tab activate the option "Run whatether user is logged on or not" (you have to enter some credentials here)
Trigger should be "On Start" (not on login)
Actions: "Execute Application `wsl -d Ubuntu`"
Uncheck everything on "Conditions" Tab.

Hope this helps someone :-)

DISCLAIMER: sorry if there wrong translations of the captions. I have a german system.

**Comment by ulayera** [comment]
This is the place where I learnt the most about wsl and the startup thing.

I wanted to keep wsl alive on windows startup, so i added a Scheduled Task to run this bat:

```
@echo off

set WSL_UTF8=1

goto main

REM Function to check WSL status
:check_status
wsl -l --running | FINDSTR Ubuntu >nul 2>&1
if %errorlevel% equ 0 (
    set "STATUS=running"
) else (
    set "STATUS=down"
)
goto :eof

REM Function to start distro
:start_distro
wsl --exec dbus-launch true
if %errorlevel% equ 0 (
    echo %DATE% %TIME% started
) else (
    echo %DATE% %TIME% failed to start (error code: %errorlevel%)
)
goto :eof

REM Main logic
:main
call :check_status
if "%STATUS%"=="running" (
    echo %DATE% %TIME% running
) else (
    echo %DATE% %TIME% down, trying to start
    call :start_distro
)

timeout /t 5 /nobreak >nul
goto main
```

run it with these params `> X:\path\to\your\logs\wsl-keepalive.log 2>&1` so the stdout gets written in a file and you can troubleshoot if necessary

**Comment by hoffor** [comment]
microsoft adding a 'solution' that doesn't actually work at all and outright ignoring user feedback, classic

it's weird.. `@reboot dbus-launch true` as a cron job doesn't keep it alive. logging into a shell and entering `dbus-launch true` works.. `@reboot bash -l -i -c dbus-launch true` does not. haven't yet found a way to automate this

no `sleep infinity` method has kept it alive for me

at the very least, `wsl --exec bash -c "dbus-launch true; echo hello"` keeps it alive for me. so basically just add that before anything you run and it should stay alive. note that this is on win11 26200 wsl2 ubuntu 24.04

**Comment by deniznehlyadyuk** [comment]
https://blog.lecoteauverdoyant.co.uk/articles/wsl-keep-alive.html

this solved my issue

**Comment by bhoppi** [comment]
> microsoft adding a 'solution' that doesn't actually work at all and outright ignoring user feedback, classic
> 
> it's weird.. `@reboot dbus-launch true` as a cron job doesn't keep it alive. logging into a shell and entering `dbus-launch true` works.. `@reboot bash -l -i -c dbus-launch true` does not. haven't yet found a way to automate this
> 
> no `sleep infinity` method has kept it alive for me
> 
> at the very least, `wsl --exec bash -c "dbus-launch true; echo hello"` keeps it alive for me. so basically just add that before anything you run and it should stay alive. note that this is on win11 26200 wsl2 ubuntu 24.04

Based on your description, here is my personal analysis:

When no user processes are running under the wsl's `/init` process (PID 2), it will terminate the WSL instance, regardless of how many other processes are running.

It's important to note that when a WSL instance starts, it launches two shell processes: one controlled by PID 2 (referred to as Shell 1), and the other invoked by the `login` command during the typical Linux system user login process (referred to as Shell 2).

Shell 1 runs first, executing the WSL system initialization commands. When it reaches the user login step, it hands over to systemd for user login, at which point Shell 2 runs.
Shell 2 executes the user login related scripts (`*profile`, `*rc`, etc.), managed by PID 1, not PID 2. After execution, the process is interrupted, and Shell 1 regains control.
Subsequent commands executed by the user in terminal all run in Shell 1.

It can be shown that:
Long-running processes managed by PID 2, including cmds executed by Shell 1, can keep the WSL instance running, for example:
- `wsl --exec bash -c "dbus-launch true; echo hello"`
- logging into a shell and entering `dbus-launch true`

Long-running processes managed by PID 1, including cmds executed by Shell 2, cannot keep the WSL instance running, for example:
- cmds (no matter `dbus-launch true`, `sleep infinity`, or others) written in `*profile` and `*rc`
- `@reboot *` cron jobs
- `loginctl enable-linger` mentioned by some other people

------

Based on the above analysis, I will present a very simple and effective solution that I have tested (this method is a modification of @gardengim 's [keepwsl](https://github.com/gardengim/keepwsl) but a little more friendly).

First, here is the content of `keepwsl.service`:
```
[Unit]
Description=Keep WSL instance alive

[Service]
Type=oneshot
ExecStart=/mnt/c/Windows/System32/wsl.exe --exec dbus-launch true
Restart=on-failure
RestartSec=5

[Install]
WantedBy=default.target
```
Put the file under `~/.config/systemd/user/`, then run below commands:
```
systemctl --user daemon-reload
systemctl --user enable --now keepwsl
```
That all, enjoy it!

**Comment by hoffor** [comment]
my personal issue with the `keepwsl.service` is that it relies on a windows drive being mounted to run its command, but i tend to keep mine dismounted because i sometimes run into issues when they're mounted (for example, one app package during installation found some windows files and really convinced itself they were the linux files it was looking for)

i suppose that could be bypassed by wrapping the wsl.exe command in a dismount-then-remount... not an elegant solution though and probably not in high demand

**Comment by gardengim** [comment]
@hoffor, I assume you could disable `appendWindowsPath` option. [https://learn.microsoft.com/en-us/windows/wsl/wsl-config#interop-settings](https://learn.microsoft.com/en-us/windows/wsl/wsl-config#interop-settings)

**Comment by hoffor** [comment]
@gardengim that is a possible solution yeah. i'm unfortunately stubborn and desire to isolate the distro's filesystem from my host's.. i think it stems from a fear i might `rm -rf /*` as a funny joke and then forget my windows drives are mounted

..yeah anyway LOL hmm i guess in that case i could send a signal to the windows environment in a different way (ssh, named pipe, creating/modifying a file in the distro and then having windows check for it, etc) and have it reactively execute the wsl dbus command. if i do anything like that i'll get back to you

**Comment by aki-k** [comment]
Of course, a new systemd service is not needed. Just start your WSL 2 instance with the wsl.exe option '--exec dbus-launch true'.

**Comment by hoffor** [comment]
### set-and-forget solution to keep a distro alive (for if you don't always keep your windows drives mounted in it):

1. in your distro: `sudo crontab -e` (choose nano if it asks)
2. paste this at the end of the file: `@reboot mkdir -p /dev/shm/wsl_ipc; chmod 777 /dev/shm/wsl_ipc; touch /dev/shm/wsl_ipc/trigger` and save the file (ctrl+x -> y -> enter)
3. back in windows, put this powershell script somewhere permanent and add it as a scheduled login task (run as admin, replace the filename with yours): `schtasks /Create /TN "wsl_keepalive" /TR "conhost powershell -WindowStyle Hidden -NonInteractive -ExecutionPolicy Bypass \"C:\wsl_keepalive.ps1\"" /SC ONLOGON /F`

wsl_keepalive.ps1:
```
$distro = "noble"
$triggerFile = "\\wsl.localhost\$distro\dev\shm\wsl_ipc\trigger"
$wslCmd = "wsl -d $distro -u root -e dbus-launch true"

while ($true)
{
	$lines = $(wsl.exe --list --running)
	foreach ($line in $lines)
	{
		$cleanedLine = $line -replace "\s\(Default\)", ""
		if ($cleanedLine -ceq $distro)
		{
			if (Test-Path $triggerFile)
			{
				Start-Process -NoNewWindow -FilePath "cmd.exe" -ArgumentList "/c $wslCmd"
				Remove-Item $triggerFile -ErrorAction SilentlyContinue
			}
			break
		}
	}
	Start-Sleep -Seconds 1
}

```

set `$distro` to your distro name (case sensitive)

you can modify the script to operate on multiple distros if you want, or you can just duplicate the script file with a different name

open Task Scheduler if you ever wanna mess with it, or to run it hidden without logging out and back in

test it:
1. `wsl -t yourdistro` to terminate it
2. `wsl -d yourdistro` to launch it and get a shell
3. `exit` to ditch it and see if windows kills it
4. if `wsl --list --running` shows it running after about 10 seconds, you're golden

**Comment by aki-k** [comment]
Holy over-engineering, Batman!

**Comment by hoffor** [comment]
> Holy over-engineering, Batman!

i agree that it shouldn't be necessary in the first place. it's only as engineered as required by the limitation we have no control over. pretty simple workaround overall but my wordiness can make it seem like more than it is

in short:
1. add a line to your crontab to create an empty file on distro boot
2. schedule a powershell script on logon that checks for that file and execute the dbus command when it finds it, then deletes it

saves you from having to exec the dbus command every time you restart your distro. for someone who does that a lot (for one reason or another) it's a life saver

**Comment by aki-k** [comment]
This is all you need:
```
wsl.exe --distribution Ubuntu-24.04 --exec dbus-launch true & wsl.exe --distribution Ubuntu-24.04
```

**Comment by aki-k** [comment]
@hoffor I received a bat signal and soon after that my e-mail box received this CMD batch script for starting and stopping a WSL 2 instance with '--exec dbus-launch true':
```
@echo off

if [%1]==[] goto usage

setlocal enableDelayedExpansion

set "_distro=Ubuntu-24.04"

for %%a in (%*) do (
    if /i "%%a"=="start" (
        set "_op=start" & goto execute
    ) else if /i "%%a"=="stop" (
        set "_op=stop" & goto execute
    ) else (
        goto usage
    )
)

:execute
if /i "%_op%"=="start" (
    wsl.exe --terminate %_distro%
    wsl.exe --distribution %_distro% --exec dbus-launch true & wsl.exe --distribution %_distro%
) else if /i "%_op%"=="stop" (
    wsl.exe --terminate %_distro%
)
endlocal
goto :EOF

:usage
echo usage: needs either a start or stop argument
endlocal
```

**Comment by hoffor** [comment]
@aki-k i boot and invoke commands from within my distro in different ways depending on what i'm doing. messing with batch scripts or remembering to manually type the dbus command (worse yet, forgetting to type it and wondering why things are broken) is what i had set out to avoid. i love my solution because it's a simple one-time setup for literally never having to think about this issue ever again. for a true brain-off solution to an issue that should have never existed, it's the end-all

**Comment by hasanfaesal** [comment]
> https://blog.lecoteauverdoyant.co.uk/articles/wsl-keep-alive.html
> 
> this solved my issue

This works for me as well, I really appreciate it. 
Thank you very much.
